### PR TITLE
gh-99108: Refactor _sha256 & _sha512 into _sha2.

### PR DESCRIPTION
**Pull request by gpshead**
This merges their code. They're backed by the same single HACL* static library, having them be a single module simplifies maintenance.

This should unbreak the wasm enscripten builds that currently fail due to linking in --whole-archive mode and the HACL* library appearing twice.

Long unnoticed error fixed: _sha512.SHA384Type was doubly assigned and was actually SHA512Type. Nobody depends on those internal names.

This is a followup to PRs #101707 and #101917.

* Issue: gh-99108

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -7197,11 +7196,8 @@ dnl By default we always compile these even when OpenSSL is available
 dnl (issue #14693). The modules are small.
 PY_STDLIB_MOD([_md5], [test "$with_builtin_md5" = yes])
 PY_STDLIB_MOD([_sha1], [test "$with_builtin_sha1" = yes])
-PY_STDLIB_MOD([_sha256],
-  [test "$with_builtin_sha256" = yes], [],
-  [-I\$(srcdir)/Modules/_hacl/include -I\$(srcdir)/Modules/_hacl/internal -D_BSD_SOURCE -D_DEFAULT_SOURCE])
-PY_STDLIB_MOD([_sha512],
-  [test "$with_builtin_sha512" = yes], [],
+PY_STDLIB_MOD([_sha2],
+  [test "$with_builtin_sha2" = yes], [],
```
Comment (erlend-aasland) [inline configure.ac:7200]: Nit: these lines can now be joined :)

File: Modules/sha2module.c [hunk 2]
```diff
@@ -0,0 +1,789 @@
...
+/* Our SHA2 implementations defer to the HACL* verified library. */
+
+#include "_hacl/Hacl_Streaming_SHA2.h"
+
+// TODO: transform .digestsize into a getter and get rid of the field?
+
+typedef struct {
+  PyObject_HEAD
+  int digestsize;
+  Hacl_Streaming_SHA2_state_sha2_256 *state;
```
Comment (erlend-aasland) [inline Modules/sha2module.c]: Nit: 4 space indent.

File: Modules/sha2module.c [hunk 3]
```diff
@@ -0,0 +1,789 @@
...
+    PyTypeObject* sha224_type;
+    PyTypeObject* sha256_type;
+    PyTypeObject* sha384_type;
+    PyTypeObject* sha512_type;
+} sha2_state;
+
+static inline sha2_state*
+sha2_get_state(PyObject *module)
+{
+    void *state = PyModule_GetState(module);
```
Comment (erlend-aasland) [inline Modules/sha2module.c]:
```suggestion
    void *state = _PyModule_GetState(module);
```

File: Modules/sha2module.c [hunk 4]
```diff
@@ -0,0 +1,789 @@
...
+    dest->digestsize = src->digestsize;
+    dest->state = Hacl_Streaming_SHA2_copy_512(src->state);
+}
+
+static SHA256object *
+newSHA224object(sha2_state *state)
+{
+    SHA256object *sha = (SHA256object *)PyObject_GC_New(SHA256object,
+                                                  state->sha224_type);
+    PyObject_GC_Track(sha);
```
Comment (erlend-aasland) [inline Modules/sha2module.c:102]: IIRC, `PyObject_GC_Track` will segfault if given a `NULL` pointer. We should do some error checking post `PyObject_GC_New`.

Ditto for the other similar functions.

File: Modules/sha2module.c [hunk 5]
```diff
@@ -0,0 +1,789 @@
...
+SHA2_traverse(PyObject *ptr, visitproc visit, void *arg)
+{
+    Py_VISIT(Py_TYPE(ptr));
+    return 0;
+}
+
+static void
+SHA256_dealloc(SHA256object *ptr)
+{
+    Hacl_Streaming_SHA2_free_256(ptr->state);
```
Comment (erlend-aasland) [inline Modules/sha2module.c:154]: Untrack first? Perhaps it does not matter.

File: Modules/sha2module.c [hunk 6]
```diff
@@ -0,0 +1,789 @@
...
+_sha2_sha256_impl(PyObject *module, PyObject *string, int usedforsecurity)
+/*[clinic end generated code: output=243c9dd289931f87 input=6249da1de607280a]*/
+{
+    Py_buffer buf;
+
+    if (string) {
+        GET_BUFFER_VIEW_OR_ERROUT(string, &buf);
+    }
+
+    sha2_state *state = PyModule_GetState(module);
```
Comment (erlend-aasland) [inline Modules/sha2module.c]:
```suggestion
    sha2_state *state = sha2_get_state(module);
```

File: Modules/sha2module.c [hunk 7]
```diff
@@ -0,0 +1,789 @@
...
+static PyObject *
+_sha2_sha224_impl(PyObject *module, PyObject *string, int usedforsecurity)
+/*[clinic end generated code: output=68191f232e4a3843 input=c42bcba47fd7d2b7]*/
+{
+    Py_buffer buf;
+    if (string) {
+        GET_BUFFER_VIEW_OR_ERROUT(string, &buf);
+    }
+
+    sha2_state *state = PyModule_GetState(module);
```
Comment (erlend-aasland) [inline Modules/sha2module.c]:
```suggestion
    sha2_state *state = sha2_get_state(module);
```

File: Modules/sha2module.c [hunk 8]
```diff
@@ -0,0 +1,789 @@
...
+    }
+    state->sha512_type = (PyTypeObject *)PyType_FromModuleAndSpec(
+        module, &sha512_type_spec, NULL);
+    if (state->sha512_type == NULL) {
+        return -1;
+    }
+
+    Py_INCREF((PyObject *)state->sha224_type);
+    if (PyModule_AddObject(module, "SHA224Type", (PyObject *)state->sha224_type) < 0) {
+        Py_DECREF((PyObject *)state->sha224_type);
```
Comment (erlend-aasland) [inline Modules/sha2module.c]: If you use `PyModule_AddType`, you can get rid of the unneeded refcount dance here. Ditto for the other types below.

**Comment by erlend-aasland** [comment]
BTW, I'm pretty sure the refleaks are related to #101908.

File: Misc/NEWS.d/next/Library/2023-02-15-01-54-06.gh-issue-99108.rjTSic.rst [hunk 9]
```diff
@@ -0,0 +1,3 @@
+The built-in extension modules for :mod:`hashlib` SHA2 algorithms, used when
+OpenSSL does provide them, now live in a single internal ``_sha2`` module
```
Comment (msprotz) [inline Misc/NEWS.d/next/Library/2023-02-15-01-54-06.gh-issue-99108.rjTSic.rst]: does not?

File: Modules/sha2module.c (hunk 5 above)
Comment (gpshead) [reply Modules/sha2module.c:154]: I don't think it matters here, the GIL is held. (long existing code)

File: configure.ac (hunk 1 above)
Comment (gpshead) [reply configure.ac:7200]: other things seem to keep them on multiple lines wen adding compiler or linker flags so I'll leave it this way.  m4 "style" is purely "be consistent with what I see nearby" for me. ;)

File: Modules/sha2module.c (hunk 8 above)
Comment (gpshead) [reply Modules/sha2module.c]: this required changing the type tp_name values to match, but that's probably a good idea anyways.

**Pull request merged**